network/peerstore/peerstore.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package peerstore
+
+import (
+	"context"
+
+	ds "github.com/ipfs/go-datastore"
+	leveldb "github.com/ipfs/go-ds-leveldb"
+	"github.com/libp2p/go-libp2p/core/peer"
+	libp2p "github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/libp2p/go-libp2p/p2p/host/peerstore/pstoreds"
+	"github.com/multiformats/go-multiaddr"
+)
+
+// PeerStore implements libp2p.Peerstore
+type PeerStore struct {
+	libp2p.Peerstore
+}
+
+func initDBStore(path string) (ds.Batching, error) {
+	store, err := leveldb.NewDatastore(path, nil)
```

Comment:
using leveldb kv store since pebbleDB is labeled experimental.

Reply:
Maybe in a future PR? As discussed, once we understand more about how the i/o usage of the peerstore is working we might want to test out different backends. We should take a look at what "experimental" means for the pebble implementation because it seems like most of our kv backed stuff is using that. Maybe there's something we can contribute upstream there as well?

Reply:
experimental probably means these work are still in progress, https://github.com/ipfs/go-ds-pebble#status

Reply:
Yeah out of that list:

> The road to maturity includes:
> * Benchmarks against go-ds-badger, go-ds-badger2, and go-ds-leveldb.
> * Exposing metrics.
> * Testing it in IPFS, libp2p, and Lotus, and characterising its behaviour, performance and footprint under real, practical workloads.

Two of those are testing and benchmarking--things which we plan to do during our integration. And the other is a feature request that maybe we could upstream if we ended up needing. So maybe it's still worth us looking at using it.

Reply:
Yeah levelDB would be 3rd DB tech. I thought that we mentioned using Sqlite? I know it's not what's used by the maintainers but is stable and is something we are using already.

On the other hand we will be using pebbleDB as @Eric-Warehime mentioned.

I prefer either of the two to levelDB for those two reasons

Reply:
changed to use pebbledb.